File: app/src/main/java/com/example/biblio/fragments/SwipeEbooksFragment.java
package com.example.biblio.fragments;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.lifecycle.Observer;
import androidx.lifecycle.ViewModelProvider;
import androidx.recyclerview.widget.LinearLayoutManager;

import com.example.biblio.adapters.EbooksAdapter;
import com.example.biblio.databinding.SwipeEbooksRvFragmentBinding;
import com.example.biblio.helpers.XFragment;
import com.example.biblio.viewmodels.EbookDetailsViewModel;
import com.example.biblio.viewmodels.SwipeEbooksViewModel;

import java.util.ArrayList;
import java.util.List;

import lrusso96.simplebiblio.core.Ebook;

public class SwipeEbooksFragment extends XFragment implements EbooksAdapter.OnItemListener {
    private final Class<? extends SwipeEbooksViewModel> mSwipeModel;
    private EbooksAdapter.OnItemListener mEbooksListener;
    private ArrayList<Ebook> mEbooks;
    private SwipeEbooksRvFragmentBinding binding;

    SwipeEbooksFragment(Class clazz, Class<? extends SwipeEbooksViewModel> model) {
        super(clazz);
        mSwipeModel = model;
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        binding = SwipeEbooksRvFragmentBinding.inflate(inflater, container, false);

        LinearLayoutManager mLayoutManager = new LinearLayoutManager(getContext());
        binding.ebooksRv.setLayoutManager(mLayoutManager);
        binding.ebooksRv.setHasFixedSize(true);
        binding.swipeContainer.setColorSchemeResources(android.R.color.holo_orange_light);

        mEbooksListener = this;
        SwipeEbooksViewModel model = new ViewModelProvider(getActivity()).get(mSwipeModel);

        binding.swipeContainer.setOnRefreshListener(model::refreshData);
        binding.swipeContainer.setRefreshing(true);
        final Observer<List<Ebook>> swipeObserver = ebooks -> {
            binding.swipeContainer.setRefreshing(true);
            logger.d("swiping");
            mEbooks = (ArrayList<Ebook>) ebooks;
            EbooksAdapter mAdapter = new EbooksAdapter(mEbooks, mEbooksListener, getContext());
            binding.ebooksRv.setAdapter(mAdapter);
            binding.swipeContainer.setRefreshing(false);
        };
        model.getEbooks().observe(getViewLifecycleOwner(), swipeObserver);
        return binding.getRoot();
    }

    @Override
    public void onItemClick(int position) {
        EbookDetailsViewModel model = new ViewModelProvider(getActivity()).get(EbookDetailsViewModel.class);
        model.setEbook(mEbooks.get(position));
        moveTo(new EbookDetailsFragment());
    }
}

File: app/src/main/java/com/example/biblio/fragments/SwipeEbooksFragment.kt
package com.example.biblio.fragments

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProvider
import androidx.recyclerview.widget.LinearLayoutManager
import com.example.biblio.adapters.EbooksAdapter
import com.example.biblio.databinding.FragmentSwipeEbooksBinding
import com.example.biblio.helpers.XFragment
import com.example.biblio.viewmodels.EbookDetailsViewModel
import com.example.biblio.viewmodels.SwipeEbooksViewModel
import lrusso96.simplebiblio.core.Ebook
import java.util.*

open class SwipeEbooksFragment internal constructor(clazz: Class<*>?, private val mSwipeModel: Class<out SwipeEbooksViewModel>) : XFragment(clazz!!), EbooksAdapter.OnItemListener {
    private lateinit var mEbooksListener: EbooksAdapter.OnItemListener
    private lateinit var mEbooks: ArrayList<Ebook>

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        super.onCreateView(inflater, container, savedInstanceState)
        val binding = FragmentSwipeEbooksBinding.inflate(inflater, container, false)
        val mLayoutManager = LinearLayoutManager(xContext)
        binding.ebooksRv.layoutManager = mLayoutManager
        binding.ebooksRv.setHasFixedSize(true)
        //binding.swipeContainer.setColorSchemeResources(android.R.color.);
        mEbooksListener = this
        val model = ViewModelProvider(activity!!)[mSwipeModel]
        binding.swipeContainer.setOnRefreshListener { model.refreshData() }
        binding.swipeContainer.isRefreshing = true
        val swipeObserver = Observer<List<Ebook>> { ebooks: List<Ebook> ->
            binding.swipeContainer.isRefreshing = true
            logger.d("swiping")
            mEbooks = ebooks as ArrayList<Ebook>
            val mAdapter = EbooksAdapter(mEbooks, mEbooksListener, xContext)
            binding.ebooksRv.adapter = mAdapter
            binding.swipeContainer.isRefreshing = false
        }
        model.getEbooks().observe(viewLifecycleOwner, swipeObserver)
        return binding.root
    }

    override fun onItemClick(position: Int) {
        val model = ViewModelProvider(activity!!).get(EbookDetailsViewModel::class.java)
        model.updateEbook(mEbooks[position])
        moveTo(EbookDetailsFragment())
    }
}